module/os/linux/zfs/zfs_uio.c

```diff
@@ -483,7 +480,12 @@ zfs_uio_free_dio_pages(zfs_uio_t *uio, zfs_uio_rw_t rw)
 	ASSERT3P(uio->uio_dio.pages, !=, NULL);
 
 #if defined(HAVE_PIN_USER_PAGES_UNLOCKED)
-	unpin_user_pages(uio->uio_dio.pages, uio->uio_dio.npages);
+	if (uio->uio_dio.pinned) {
+		unpin_user_pages(uio->uio_dio.pages, uio->uio_dio.npages);
+		vmem_free(uio->uio_dio.pages,
+		    uio->uio_dio.npages * sizeof (struct page *));
+		return;
+	}
 #else
 	for (long i = 0; i < uio->uio_dio.npages; i++) {
```

Comment:
I have feeling that instead of `#else` here should be `#endif`, so that in case of `!pinned` we fall back to `put_pages()` below.

Reply:
Yep, that as a complete brain fart on my part. I updated this.